ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/encoders/TomlAbstractEncoder.kt

```diff
@@ -0,0 +1,388 @@
+        parent.appendValue(value)
+    }
+
+    override fun encodeBoolean(value: Boolean) {
+        if (!encodeAsKey(value, "Boolean")) {
+            appendValue(TomlBoolean(value, elementIndex))
+        }
+    }
+
+    override fun encodeDouble(value: Double) {
+        if (!encodeAsKey(value, "Double")) {
+            appendValue(TomlDouble(value, elementIndex))
+        }
+    }
+
+    override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) {
+        encodeString(enumDescriptor.getElementName(index))
+    }
+
+    override fun encodeLong(value: Long) {
+        if (attributes.intRepresentation != DECIMAL) {
+            throw UnsupportedEncodingFeatureException(
+                "Non-decimal integer representation is not yet supported."
+            )
+        }
+
+        if (!encodeAsKey(value, "Long")) {
+            appendValue(TomlLong(value, elementIndex))
+        }
+    }
+
+    override fun encodeNull() {
+        appendValue(TomlNull(elementIndex))
+    }
+
+    override fun encodeString(value: String) {
+        if (attributes.isMultiline) {
+            throw UnsupportedEncodingFeatureException(
+                "Multiline strings are not yet supported."
+            )
+        }
+
+        if (!encodeAsKey(value)) {
+            appendValue(
+                if (attributes.isLiteral) {
+                    TomlLiteralString(value as Any, elementIndex)
+                } else {
+                    TomlBasicString(value as Any, elementIndex)
+                }
+            )
+        }
+    }
+
+    override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
+        when (val desc = serializer.descriptor) {
+            instantDescriptor,
+            localDateTimeDescriptor,
+            localDateDescriptor -> if (!encodeAsKey(value as Any, desc.serialName)) {
+                appendValue(TomlDateTime(value as Any, elementIndex))
+            }
+            else -> when (val kind = desc.kind) {
+                is StructureKind,
+                is PolymorphicKind -> if (!encodeAsKey(value as Any, desc.serialName)) {
+                    val encoder = encodeStructure(kind)
+
+                    serializer.serialize(encoder, value)
+
+                    elementIndex = encoder.elementIndex
+
+                    attributes.reset()
+                }
+                else -> super.encodeSerializableValue(serializer, value)
+            }
+        }
+    }
+
+    override fun encodeByte(value: Byte): Nothing = invalidType("Byte", "Long", value)
+    override fun encodeShort(value: Short): Nothing = invalidType("Short", "Long", value)
+    override fun encodeInt(value: Int): Nothing = invalidType("Int", "Long", value)
+    override fun encodeFloat(value: Float): Nothing = invalidType("Float", "Double", value)
+    override fun encodeChar(value: Char): Nothing = invalidType("Char", "String", value)
+
+    // Todo: Do we really want to make these invalid?
```

Comment:
yeah, I also started thinking that following toml specification in a so straight-forward way can be a problem

Reply:
May be later? Can you please create a separate issue? Let’s not take too much in one PR :)